### DISCUSSION TITLE
Building a browser.

### DISCUSSION THREAD

**Discussion by MBoui** [tauri-apps/tauri#4219] [General]
Hi the community,

I like Tauri, and i want to use him everywhere! I think to start a new project which will be a little web browser.
I am really curious about what do you think to start this kind of project with Tauri? Do you see any limitation or concern for this kind of project?

**Comment by JonasKruckenberg** [comment]
Without wanting to hurt your feelings: Don't.

[here's a bit about why](https://github.com/tauri-apps/tauri/issues/2975#issuecomment-1019945044)

Essentially building a browser is **very different** from building a webview UI where the source is trusted.
Enabling this on our end will require *significant* effort to get right, something that we cannot handle.

Building a browser also requires a lot more effort on your end than you might imagine. A browser is almost on-par with operating systems when it comes to complexity.

Interestingly enough electron and Tauri are on the same page here. Executing untrusted content in either framework is incredibly insecure and **highly**, **highly**, **highly** discouraged.

Edit: If you want to build a Browser, you're much better off doing what Brave or Edge did and use the CEF (chromium embedded framework) directly

**Reply by Jocabin** [reply]
I like your answer, but what if I don't want to use chromium but tauri engine ?

**Reply by AZProductions** [reply]
> I like your answer, but what if I don't want to use chromium but tauri engine ?
> 

Wouldn't it then be Webview2 on Windows, and Webview2 is a highly modified fork of chromium.

**Reply by nileshtrivedi** [reply]
FWIW, the issue @JonasKruckenberg linked above, #2975 , has been fixed a few months ago.

**Reply by KaKi87** [reply]
> electron and Tauri are on the same page here.

Where do Electron devs say that ?

**Reply by nileshtrivedi** [reply]
FWIW, [BonBon Browser](https://github.com/BonBon-exchange/bonbon-browser) project is built on top of Electron.

**Reply by wyntxr** [reply]
chromium is bloat, just use webkit or servo

**Reply by KaKi87** [reply]
Chromium is full-featured. WebKit and Servo are way late on a lot of things.

**Comment by autonome** [comment]
> Edit: If you want to build a Browser, you're much better off doing what Brave or Edge did and use the CEF (chromium embedded framework) directly

Brave and Edge both build on forked Chromium, but neither use CEF.

**Reply by AZProductions** [reply]
Well it's even easier to use cef when building your own browser

**Reply by Notliam99** [reply]
Firefox . . . Would it be possible i mean they have no equivalent to electron witch i find weird but it could be a option but its a very complex project i bet

**Reply by AZProductions** [reply]
> Firefox . . . Would it be possible i mean they have no equivalent to electron witch i find weird but it could be a option but its a very complex project i bet 

Sadly Firefox does not have a webview for desktop platforms. Which makes it virtually impossible, because you'd need to fork it and maintain it.

**Comment by Zack-Xb** [comment]
Is it possible to just load a website that is specified to load via tauri?

**Reply by mahadi-hossain** [reply]
You can build but you know turi does not provide so many functionalities. I always struggle with dilatory launching speed of established browsers. So I decided to build a very light and blazing first one by using tauri, in this case security does not matter for my personal use

**Comment by Abhishek-Jaiswar** [comment]
Did someone created yet?

**Comment by nileshtrivedi** [comment]
I built a very rough POC in case anybody is interested. I don't want to invest too much time into it until multiwebview becomes a stable feature or unless some one pitches in for help.

![image](https://github.com/user-attachments/assets/3c2c538d-8fb5-427a-a308-c08b394fbc2d)

EDIT: I have pushed my code example in the `osdemo` branch here: https://github.com/nileshtrivedi/tauri/tree/osdemo

This is added as a new example [here](https://github.com/nileshtrivedi/tauri/tree/osdemo/examples/osdemo). It can be run with from the top directory of the repo with `cargo run --example osdemo --features unstable`

It's rough because the main webview is a single webpage index.html using TailwindCSS and DaisyUI via CDN with imperative Javascript code for adding/listing/removing webviews, whereas for a real-app I'd like to use some kind of component framework like Svelte or Preact.

Also, the webview size and placement is currently random. Also I have not yet figured out how to add window border and decoration because the webviews positioning is done in Rust but without any border. DaisyUI has a mock browser component but the size/position of webviews in Rust is with respect to the whole window/screen which is tricky to communicate and handle inside the main webview.

**Reply by KaKi87** [reply]
Hi,

> I built a very rough POC in case anybody is interested

Could you please share the code in a repo ?

> until multiwebview becomes a stable feature

The feature was shipped in v2.0.0, right ?

**Reply by autonome** [reply]
Folks in the User & Agents community are poking at similar directions, you'd likely get interest if there was a repo w/ example code.

**Reply by nileshtrivedi** [reply]
I have pushed my code example in the `osdemo` branch here: https://github.com/nileshtrivedi/tauri/tree/osdemo

This is added as a new example [here](https://github.com/nileshtrivedi/tauri/tree/osdemo/examples/osdemo). It can be run with from the top directory of the repo with `cargo run --example osdemo --features unstable`

It's rough because the main webview is a single webpage index.html using TailwindCSS and DaisyUI via CDN with imperative Javascript code for adding/listing/removing webviews, whereas for a real-app I'd like to use some kind of component framework like Svelte or Preact.

Also, the webview size and placement is currently random. Also I have not yet figured out how to add window border and decoration because the webviews positioning is done in Rust but without any border. DaisyUI has a mock browser component but the size/position of webviews in Rust is with respect to the whole window/screen which is tricky to communicate and handle inside the main webview.

**Comment by LuanderFarias** [comment]
for those coming here later, 3 months ago tauri made the function. you can now create more webviews :D

**Reply by KaKi87** [reply]
Has anyone done it ?

**Reply by LuanderFarias** [reply]
> Has anyone done it ?

as far as i've seen, no. i'm making a browser and tought about using tauri but decided to go with chromium, but i'm sure it's possible. also if you don't want webview there's "cef-rs", that can make you a more complete browser

**Reply by KaKi87** [reply]
I'm actually a JS person, so I would indeed use Tauri for its JS API rather than CEF.

**Reply by LuanderFarias** [reply]
> I'm actually a JS person, so I would indeed use Tauri for its JS API rather than CEF.

Fair enough, then i'm pretty sure you can do an amazing browser if you want :)

**Comment by fr1g** [comment]
I suppose this is something necessary... 
have you heard a browser for android named Via ?
this stuff is directly using android webview. this kind of solution is quite saving space, due to the OS integrating webview is aimed to save storage, reducing unnecessary duplicating modules (you all already know so :)

so why not remove edge itself but keep webview2 to create a lighter frame to implement a small yet functional browser?

**Reply by rc-zb** [reply]
What do you think about [cosmic-utils/web-apps](https://github.com/cosmic-utils/web-apps)? It uses Tau plus Wry, but not Tauri. However, I haven't seen more restriction for security in raw Wry than Tauri.
